lib/ClusterShell/Engine/Poll.py

```diff
@@ -109,7 +109,12 @@ def runloop(self, timeout):
                     timeo = timeout
 
                 self._current_loopcnt += 1
-                evlist = self.polling.poll(timeo * 1000.0 + 1.0)
+
+                if timeo < 0:
+                    poll_timeo = -1
+                else:
+                    poll_timeo = timeo * 1000.0 + 1.0
```

Comment:
I think the only code path to get a buggy value for this math is when timeo == -1 above (line 108).
The only place where this math should be done is when timeo is positive, so on line 107. 

I think you should simply move this math on line 107 and simply pass timeo to `poll()`

By the way, EPoll should have the same logic. I do not know if it has the same bug, but it is a good idea to also fix it the same way.

Reply:
@degremont thanks for the quick review. The reason why I did it there is that `timeo = timeout` is also set on line 109 if `timeo == -1` and `timeout` is a parameter of the `runloop` method, that could potentially be negative. I didn't want to make assumptions related to the result of the `nextfire_delay()` call and the possibility to have a negative timeout passed, so erring on the safe side.
If you tell me that is safe to move the math logic to line 107 I'll do it.

Regarding `EPoll`, that's a Linux-only system call and from a quick test although the man page for `epoll_wait` states that `-1` should be specified, seems that it accept any negative value like the `poll` system call. Anyway is indeed safer to pass exactly `-1` to wait indefinitely. I'll include this change in the PR.

Reply:
@volans- I've checked values returned by `nextfire_delay()` and possible values for `timeout`. 
Actually my proposal was wrong if `nextfire_delay()` returns -1 and `timeout` is positive.
After testing various possibilities of improvements, I could not find a nice way to write this.

I could not get something better than you. Please just update EPoll too.

@thiell You have a better knowledge than me of this part of the code. Maybe you will have a suggestion?

Reply:
@degremont ok, PR updated.
I was also wondering if the case of adding 1 millisecond was to avoid to pass `0` as timeout to the `poll()` methods if/when `nextfire_delay()` returns `0`, given that `0` has a special meaning in the `poll()` methods. As of the current implementation `0` will never be passed as `timeout`, just checking that is the intended behaviour given that we're at it.

Reply:
@volans- please get rid of the added millisecond, I don't recall any good reason to keep it. Then `0` will be passed to `poll()` or `epoll_wait()` when `self.timerq.nextfire_delay()` returns `0`, which may happen in case of laggy timers, but that's fine, the call will be non-blocking and should trigger any pending events + process pending timers.

The patch looks good to me, and yes, please update EPoll.py too. Thanks much!

Reply:
@degremont @thiell done.